lib/rubocop/cop/utils/regexp_ranges.rb

```diff
@@ -6,27 +6,41 @@ module Utils
       # Helper to abstract complexity of building range pairs
       # with octal escape reconstruction (needed for regexp_parser < 2.7).
       class RegexpRanges
-        attr_reader :compound_token, :root
+        attr_reader :root
 
         def initialize(root)
           @root = root
-          @compound_token = []
+          @populated = false
+        end
+
+        def compound_token
+          populate_all unless @populated
+
+          @compound_token
         end
 
         def pairs
-          unless @pairs
-            @pairs = []
-            populate(root)
-          end
+          populate_all unless @populated
+
+          @pairs
+        end
+
+        private
+
+        def populate_all
+          @compound_token = []
+          @pairs = []
+
+          populate(@root)
 
           # If either bound is a compound the first one is an escape
           # and that's all we need to work with.
           # If there are any cops that wanted to operate on the compound
           # expression we could wrap it with a facade class.
-          @pairs.map { |pair| pair.map(&:first) }
-        end
+          @pairs = @pairs.map { |pair| pair.map(&:first) }
```

Comment:
Using `map!` would be simple.

```suggestion
          @pairs.map! { |pair| pair.map(&:first) }
```